### PR TITLE
[Update distribution.py] Add probability evaluation

### PR DESCRIPTION
**Pull request by zuoxingdong**
Useful for e.g. computing loss for Mixture Density Networks.

### CODE REVIEW

**Comment by apaszke** [comment]
cc: @fritzo. I think we might want to avoid implementing that as `exp` of `log_prob`, because it might be numerically unstable.

**Comment by fritzo** [comment]
@zuoxingdong can you add some override implementations in this PR to provide motivation? Without numerical motivation this seems like interface creep.

@apaszke I agree it's usually a bad idea to use raw probabilities in ML due to overflow issues. I like that we currently provide `.log_prob()` and not `.prob()` as a slight nudge to beginning ML practitioners to suggest they keep probabilites in log units rather than raw units.

**Comment by zuoxingdong** [comment]
@fritzo For example, in mixture density networks, the loss function is defined as 
`L = -1/N*\sum_{i=1}^{N}ln( \sum_{k=1}^{K} \pi_k(x_i) N(\mu_k(x_i), \sigma_k(x_i)) )`
where the Gaussian probability for each input dimension is calculated and multiplied with another vector `\pi` for mixing coefficient, and then summed up together, before finally fit into `log` operation. So I feel it is nontrivial to use current `log_prob` API to do it neatly.

**Comment by zuoxingdong** [comment]
@eb8680 Just to make sure if I got it correctly, the `log_sum_exp` trick helps in the Gaussian case (exponential family). When we need to have different density components, it also works like that ?

**Comment by eb8680** [comment]
@zuoxingdong it's not specific to exponential families; see e.g. [this blog post](https://hips.seas.harvard.edu/blog/2013/01/09/computing-log-sum-exp/).  Your example would become:
```python
def mdn_loss(pi, mu, sigma, target):
   log_prob = pi*Normal(mu, sigma).log_prob(target)
   nll_loss = -log_sum_exp(log_prob, dim=1)
   
   return nll_loss.mean()
```

**Comment by fritzo** [comment]
> moved upstream

Consider using `torch.distributions.utils.log_sum_exp()` https://github.com/pytorch/pytorch/blob/master/torch/distributions/utils.py#L116

**Comment by zuoxingdong** [comment]
@eb8680 @fritzo Thanks for the help. Now it works well. With this trick, I also don' see other demands for raw probabilities, so if you think it is okay, I would like to close this PR.

**Comment by alok** [comment]
TRPO uses raw probabilities, plus PyTorch is useful for more than just ML, since it has a nice numpy interface over gpu acceleration.
